src/libAtomVM/ets.c

```diff
@@ -271,39 +265,79 @@ EtsErrorCode ets_insert(term ref, term entry, Context *ctx)
 
     Heap *heap = malloc(sizeof(Heap));
     if (IS_NULL_PTR(heap)) {
-        SMP_UNLOCK(ets_table);
         return EtsAllocationFailure;
     }
     size_t size = (size_t) memory_estimate_usage(entry);
     if (memory_init_heap(heap, size) != MEMORY_GC_OK) {
         free(heap);
-        SMP_UNLOCK(ets_table);
         return EtsAllocationFailure;
     }
 
     term new_entry = memory_copy_term_tree(heap, entry);
     term key = term_get_tuple_element(new_entry, (int) ets_table->keypos);
 
-    EtsErrorCode ret = EtsOk;
+    EtsErrorCode result = EtsOk;
     EtsHashtableErrorCode res = ets_hashtable_insert(ets_table->hashtable, key, new_entry, EtsHashtableAllowOverwrite, heap, ctx->global);
     if (UNLIKELY(res != EtsHashtableOk)) {
-        ret = EtsAllocationFailure;
+        result = EtsAllocationFailure;
     }
 
-    SMP_UNLOCK(ets_table);
+    return result;
+}
 
-    return ret;
+static EtsErrorCode ets_table_insert_list(struct EtsTable *ets_table, term list, Context *ctx)
+{
+    term list1 = list;
+
+    while (term_is_nonempty_list(list1)) {
+        term tuple = term_get_list_head(list1);
+        list1 = term_get_list_tail(list1);
+        if (!term_is_tuple(tuple) || term_get_tuple_arity(tuple) < 1) {
+            SMP_UNLOCK(ets_table);
+            return EtsBadEntry;
+        }
+    }
+    if (!term_is_nil(list1)) {
+        return EtsBadEntry;
+    }
+
+    while (term_is_nonempty_list(list)) {
+        term tuple = term_get_list_head(list);
+        EtsErrorCode result = ets_table_insert(ets_table, tuple, ctx);
+        if (UNLIKELY(result != EtsOk)) {
+            AVM_ABORT(); // Abort because operation might not be atomic.
+        }
+
+        list = term_get_list_tail(list);
+    }
+
+    return EtsOk;
 }
 
-EtsErrorCode ets_lookup(term ref, term key, term *ret, Context *ctx)
+EtsErrorCode ets_insert(term name_or_ref, term entry, Context *ctx)
 {
-    struct EtsTable *ets_table = term_is_atom(ref) ? ets_get_table_by_name(&ctx->global->ets, ref, TableAccessRead) : ets_get_table_by_ref(&ctx->global->ets, term_to_ref_ticks(ref), TableAccessRead);
+    struct EtsTable *ets_table = term_is_atom(name_or_ref) ? ets_get_table_by_name(&ctx->global->ets, name_or_ref, TableAccessWrite) : ets_get_table_by_ref(&ctx->global->ets, term_to_ref_ticks(name_or_ref), TableAccessWrite);
     if (ets_table == NULL) {
         return EtsTableNotFound;
     }
+    EtsErrorCode result;
+
+    if (term_is_tuple(entry) && term_get_tuple_arity(entry) > 0) {
```

Comment:
Nit: keep result glued to if-else chain to improve clarity. Also, arity check isn't needed, it's checked deeper. IMO we should do it in `nifs.c` though and ensure that only keypos is checked (we'd have to iterate through list of entries in case of list of tuples which is unfortunate but keeps concerns separated).